interpreter/luajit/proto.go

```diff
@@ -0,0 +1,362 @@
+	"github.com/open-telemetry/opentelemetry-ebpf-profiler/remotememory"
+)
+
+const (
+	sizeofGCstr   = 24
+	sizeofGCproto = 104
+	stringGCType  = 4
+	//https://github.com/openresty/luajit2/blob/7952882d/src/lj_def.h#L66
+	byteCodeMax = 1 << 26
+)
+
+type GCobj struct {
+	_   uint64 // nextgc
+	_   byte   // marked
+	gct byte
+}
+
+// GCproto minus first 8 bytes
+// https://github.com/openresty/luajit2/blob/7952882d/src/lj_obj.h#L372
+// All the pointers (except chunkname) are pointers to extra space at the end of the GCproto object
+// so we could try to be clever and read the whole thing at once if we needed to reduce remotememory
+// traffic.
+type protoRaw struct {
+	// nextgc uint64    /*      0      |       8 */
+	_      byte   /*      8      |       1 */
+	_      byte   /*      9      |       1 */
+	_      byte   /*     10      |       1 */
+	_      byte   /*     11      |       1 */
+	sizebc uint32 /*     12      |       4 */
+	_      uint32 /*     16      |       4 */
+	/* XXX  4-byte hole      */
+	_         uint64        /*     24      |       8 */
+	k         libpf.Address /*     32      |       8 */
+	_         uint64        /*     40      |       8 */
+	sizekgc   uint32        /*     48      |       4 */
+	_         uint32        /*     52      |       4 */
+	sizept    uint32        /*     56      |       4 */
+	sizeuv    uint8         /*     60      |       1 */
+	_         uint8         /*     61      |       1 */
+	_         uint16        /*     62      |       2 */
+	chunkname libpf.Address /*     64      |       8 */
+	firstline uint32        /*     72      |       4 */
+	numline   uint32        /*     76      |       4 */
+	lineinfo  libpf.Address /*     80      |       8 */
+	uvinfo    libpf.Address /*     88      |       8 */
+	varinfo   libpf.Address /*     96      |       8 */
+}
+
+type proto struct {
+	protoRaw
+	ptAddr       libpf.Address
+	name         string
+	bc           []uint32
+	lineinfo8    []uint8
+	lineinfo16   []uint16
+	lineinfo32   []uint32
+	upvalueNames []string
+	varinforaw   []byte
+	constants    []string
+}
+
+func newProto(rm remotememory.RemoteMemory, pt libpf.Address) (*proto, error) {
+	p := &proto{ptAddr: pt}
+	if err := rm.Read(pt+8, libpf.SliceFrom(&p.protoRaw)); err != nil {
+		return nil, err
+	}
+
+	// reading memory from a remote process is always dicey, validate
+	// we're looking at a GCproto object by checking that the
+	end := pt + libpf.Address(p.sizept)
+	bad := func(addr libpf.Address) bool {
+		return addr != 0 && (addr < pt || addr >= end)
+	}
+	if bad(p.lineinfo) || bad(p.uvinfo) || bad(p.varinfo) {
+		return nil, errors.New("invalid GCproto object")
+	}
+
+	// string data is stored after the GCstr object
+	p.name = rm.String(p.chunkname + sizeofGCstr)
+	if !utf8.ValidString(p.name) {
+		return nil, errors.New("invalid chunkname string")
+	}
+	// If p.name == "" should we set it to something that isn't a valid lua function name?
+	// TODO: what does lua do for anonymous functions?
```

Comment:
Unfortunately `rm.String` returns `""` in the error case, which doesn't let us distinguish between a legitimate empty string, and an error reading memory.

Reply:
I don't think a chunkname can be an empty string.  Its never null and it always gets set so I think this is fine.  I'll investigate and remove the TODO.

Reply:
Oh wait chunkname isn't about the function name its about the file name/source.  Its never empty.